### PR TITLE
feat: update handler map when field descriptor hashcode changes

### PR DESCRIPTION
**Pull request by Meghajit**
## Issue

In the dagger repository, the  TypeHandlerFactory class maintains a static hash map, which memoizes which handler to use for serialization/deserialization for every proto full name. The type handler objects are created once using the field descriptor and saved as the value of this hash map.

The hypothesis was that this map was being persisted in the memory of a task manager, even when the job terminates or restarts. As a result the handler objects get reused and at the time of serialization when this method is called in the stacktrace:

```
at com.google.protobuf.DynamicMessage$Builder.verifyContainingType(DynamicMessage.java:619)
```

the equals comparison fails and hence serialization also fails.

## Fix

The fix is then to update the handler corresponding to the proto full name in the hash map,  whenever the field descriptor hashcode is a new one.

### CODE REVIEW

File: dagger-common/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 1]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (Meghajit) [inline dagger-common/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: `Descriptors.FieldDescriptor` is a final class. In order to mock it, we need to enable this mockito extension

Ref: https://www.baeldung.com/mockito-final#configure-mocktio

Comment (Meghajit) [reply dagger-common/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker]: Using powermockito instead

File: dagger-common/build.gradle [hunk 2]
```diff
@@ -80,6 +80,8 @@ dependencies {
     testImplementation 'com.google.protobuf:protobuf-java:3.5.0'
     testImplementation 'com.google.protobuf:protobuf-java-util:3.5.0'
     testImplementation 'org.grpcmock:grpcmock-junit5:0.5.0'
+    testImplementation 'org.powermock:powermock-module-junit4:2.0.0-beta.5'
+    testImplementation 'org.powermock:powermock-api-mockito2:2.0.0-beta.5'
```
Comment (prakharmathur82) [inline dagger-common/build.gradle:84]: let's not use powermock if possible, it messes up the classpaths

File: dagger-common/src/test/java/com/gotocompany/dagger/common/serde/typehandler/TypeHandlerFactoryTest.java [hunk 3]
```diff
@@ -12,19 +16,47 @@
...
+        Descriptors.FieldDescriptor mapFieldDescriptor = TestBookingLogMessage.getDescriptor().findFieldByName("metadata");
+        TypeHandler typeHandler1 = TypeHandlerFactory.getTypeHandler(mapFieldDescriptor);
+        TypeHandler typeHandler2 = TypeHandlerFactory.getTypeHandler(mapFieldDescriptor);
+        assertEquals(typeHandler1, typeHandler2);
+    }
+
+    @Test
+    public void shouldReturnDifferentCopiesOfHandlerObjectWhenFieldDescriptorFullNameIsSameButHashCodeIsDifferent() {
+        Descriptors.FieldDescriptor mapFieldDescriptor1 = TestBookingLogMessage.getDescriptor().findFieldByName("metadata");
+        Descriptors.FieldDescriptor mapFieldDescriptor2 = PowerMockito.mock(Descriptors.FieldDescriptor.class);
```
Comment (prakharmathur82) [inline dagger-common/src/test/java/com/gotocompany/dagger/common/serde/typehandler/TypeHandlerFactoryTest.java:51]: if you again create this, how the hashcode will be same?

Comment (Meghajit) [reply dagger-common/src/test/java/com/gotocompany/dagger/common/serde/typehandler/TypeHandlerFactoryTest.java]: Changed to use protobuf libraries

File: dagger-common/build.gradle (hunk 2 above)
Comment (Meghajit) [reply dagger-common/build.gradle]: Removed

**Review by sumitaich1998 — APPROVED** [review]

**Pull request merged**